File: src/FASTCAMPUS/ALGO/D14/Main.java
package FASTCAMPUS.ALGO.D14;

import java.util.ArrayList;

class Main {
  public static void main(String[] args) {
    NQueen nQueen = new NQueen();
    nQueen.search(4, 0, new ArrayList<Integer>());

  }
}

class NQueen{
  boolean check(ArrayList<Integer>candidate,Integer currentColumn){
    int currentRow = candidate.size();
    for(int i = 0; i < currentRow; i ++){
      if(candidate.get(i) == currentColumn || (Math.abs(candidate.get(i) - currentColumn) == (currentRow - i))){
        return false;
      }
    }
    return true;
  }
  void search(Integer N, Integer currentRow, ArrayList<Integer>currentCandidate){
    if(N == currentRow){
      System.out.println(currentCandidate);
      return;
    }
    for( int i = 0; i < N; i ++){
      if(check(currentCandidate, i) == true){
        currentCandidate.add(i);
        this.search(N, currentRow + 1, currentCandidate);
        currentCandidate.remove(currentCandidate.size() - 1);
      }
    }
  }
}

File: src/FASTCAMPUS/ALGO/D14/Main.js


const validate = function(candidate , currentCol){
  // currentRow : 퀸을 위치시킬 행
  // currentCol : 퀸을 위치시킬 열
  let currentRow = candidate.length;
  for(let i = 0; i < currentRow; i ++){
    //각행에 위치한 퀸의 열
    let queenCol = candidate[i];
    if((queenCol === currentCol) || Math.abs(currentRow - i) === Math.abs(currentCol - queenCol)){
      return false;
    }
  }
  return true;
}
const nqueen = function(N, depth, candidate,){
  if(depth === N){
    console.log(candidate);
    return;
  }
  for(let i = 0; i < N ; i ++){
    if(validate(candidate, i)){
      candidate.push(i);
      nqueen(N, depth + 1, candidate);
      candidate.pop();
    }
  }
}
nqueen(4,0,[]);